task_processing/plugins/mesos/execution_framework.py

```diff
@@ -55,17 +57,28 @@ def __init__(
             checkpoint=True,
             role=self.role
         )
-        self.task_queue = Queue(max_task_queue_size)
-        self.task_update_queue = Queue(max_task_queue_size)
+
+        self.task_queue = make_thread_safe_struct(
```

Comment:
Immutable data structures don't help with mutable races (e.g. the queues, that's what locking is for), but they do help limit the mutability within those mutables stores which limits races.

For example, imagine that you have a queue of dicts, and someone peeks at the top dict and modifies the 'done' key. In the pyrsistent world that can't happen, so you don't have to worry about other threads of execution messing with your state. You must instead consume the item from the queue and add a new item back on, which is as previously mentioned thread safe (because of the Queue internal locking).

We definitely want to move to a coroutine model IMO, which is part of the reason we're moving all the things to Python3 (and the faster we use gevent or asyncio the better IMO), but honestly you end up with the same problems since Python threads can't actually run concurrently (same as green threads or coroutines); you still have to lock critical sections when you have coroutines.